bin/faucet/src/client.rs

```diff
@@ -0,0 +1,353 @@
+use std::{cell::RefCell, rc::Rc, time::Duration};
+
+use miden_lib::{
+    accounts::faucets::create_basic_fungible_faucet, notes::create_p2id_note, AuthScheme,
+};
+use miden_node_proto::generated::{
+    requests::{GetBlockHeaderByNumberRequest, SubmitProvenTransactionRequest},
+    rpc::api_client::ApiClient,
+};
+use miden_objects::{
+    accounts::{Account, AccountDelta, AccountId, AccountStorageType, AuthSecretKey},
+    assembly::{ModuleAst, ProgramAst},
+    assets::{FungibleAsset, TokenSymbol},
+    crypto::{
+        dsa::rpo_falcon512::{self, Polynomial, SecretKey},
+        merkle::{MmrPeaks, PartialMmr},
+        rand::RpoRandomCoin,
+    },
+    notes::{Note, NoteId, NoteType},
+    transaction::{ChainMmr, ExecutedTransaction, InputNotes, TransactionArgs},
+    vm::AdviceMap,
+    BlockHeader, Felt, Word,
+};
+use miden_tx::{
+    utils::Serializable, AuthenticationError, DataStore, DataStoreError, ProvingOptions,
+    TransactionAuthenticator, TransactionExecutor, TransactionInputs, TransactionProver,
+};
+use rand::{thread_rng, Rng};
+use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};
+use tonic::transport::Channel;
+
+use crate::{config::FaucetConfig, errors::FaucetError};
+
+pub const DISTRIBUTE_FUNGIBLE_ASSET_SCRIPT: &str =
+    include_str!("transaction_scripts/distribute_fungible_asset.masm");
+
+pub struct FaucetClient {
+    rpc_api: ApiClient<Channel>,
+    executor: TransactionExecutor<FaucetDataStore, FaucetAuthenticator>,
+    data_store: FaucetDataStore,
+    id: AccountId,
+    rng: RpoRandomCoin,
+}
+
+unsafe impl Send for FaucetClient {}
+unsafe impl Sync for FaucetClient {}
+
+impl FaucetClient {
+    pub async fn new(config: FaucetConfig) -> Result<Self, FaucetError> {
+        let (rpc_api, root_block_header, root_chain_mmr) =
+            initial_connection(config.clone()).await?;
```

Comment:
nit: The name for this function is a bit misleading, maybe something like `initialize_faucet_client`? Since it requests the genesis block, initliazes the tonic client, etc.

Reply:
Done.